### PR TITLE
Feat 1

### PR DESCRIPTION
**Pull request by younghoondoodoom**
### **변경 사항**

**AS-IS**

**TO-BE**
1. 회원 가입
    - 회원가입 페이지 접근
    - 이메일, 비밀번호, 비밀번호 확인, 닉네임 입력을 통한 회원가입
    - 입력 값 유효성 검사
        - 이메일 형식, 비밀번호 기준(영어, 숫자 무조건 포함한 8~20글자, 특수기호 포함 가능), 비밀번호 일치, 모든 값은 빈칸이면 안됨
        - 유효성 검증에 실패하면 에러페이지 반환(에러코드, 에러메시지 포함)
    - 이메일이 이미 있는 값이라면 에러페이지 반환(에러코드, 에러메시지 포함)
    - default status는 ACTIVE, default isAdmin은 false
    - 회원 생성일자와 수정일자 포함.
2. 로그인/ 로그아웃
    - session을 통한 로그인 전략 선택
    - 이메일, 비밀번호를 통한 로그인
    - 로그인 실패 시 로그인 페이지로 redirect
    - 해당 계정 정지 시 에러페이지 반환(에러코드, 에러미시지 포함)
 

### **테스트**

- [x]  테스트 코드
- [x]  API 테스트

### CODE REVIEW

**Comment by younghoondoodoom** [comment]
|File|Coverage [73.45%]|:x:|
|:-|:-:|:-:|
|[MemberException.java](https://github.com/younghoondoodoom/peer-market/blob/d3842bbf7000f19764c0bfd1a40eb7e5fbb4f629/src%2Fmain%2Fjava%2Fdoodoom%2Fproject%2Fpeermarket%2Fexception%2FMemberException.java)|100%|:green_apple:|
|[MemberServiceImpl.java](https://github.com/younghoondoodoom/peer-market/blob/d3842bbf7000f19764c0bfd1a40eb7e5fbb4f629/src%2Fmain%2Fjava%2Fdoodoom%2Fproject%2Fpeermarket%2Fservice%2Fmember%2FMemberServiceImpl.java)|100%|:green_apple:|
|[ErrorCode.java](https://github.com/younghoondoodoom/peer-market/blob/d3842bbf7000f19764c0bfd1a40eb7e5fbb4f629/src%2Fmain%2Fjava%2Fdoodoom%2Fproject%2Fpeermarket%2Ftype%2FErrorCode.java)|100%|:green_apple:|
|[MemberStatus.java](https://github.com/younghoondoodoom/peer-market/blob/d3842bbf7000f19764c0bfd1a40eb7e5fbb4f629/src%2Fmain%2Fjava%2Fdoodoom%2Fproject%2Fpeermarket%2Ftype%2FMemberStatus.java)|100%|:green_apple:|
|[BaseTimeEntity.java](https://github.com/younghoondoodoom/peer-market/blob/d3842bbf7000f19764c0bfd1a40eb7e5fbb4f629/src%2Fmain%2Fjava%2Fdoodoom%2Fproject%2Fpeermarket%2Fdomain%2Fbase%2FBaseTimeEntity.java)|100%|:green_apple:|
|[SecurityConfig.java](https://github.com/younghoondoodoom/peer-market/blob/d3842bbf7000f19764c0bfd1a40eb7e5fbb4f629/src%2Fmain%2Fjava%2Fdoodoom%2Fproject%2Fpeermarket%2Fconfiguration%2FSecurityConfig.java)|100%|:green_apple:|
|[JpaAuditingConfig.java](https://github.com/younghoondoodoom/peer-market/blob/d3842bbf7000f19764c0bfd1a40eb7e5fbb4f629/src%2Fmain%2Fjava%2Fdoodoom%2Fproject%2Fpeermarket%2Fconfiguration%2FJpaAuditingConfig.java)|100%|:green_apple:|
|[CustomAuthenticationFailureHandler.java](https://github.com/younghoondoodoom/peer-market/blob/d3842bbf7000f19764c0bfd1a40eb7e5fbb4f629/src%2Fmain%2Fjava%2Fdoodoom%2Fproject%2Fpeermarket%2Fconfiguration%2FCustomAuthenticationFailureHandler.java)|100%|:green_apple:|
|[MemberController.java](https://github.com/younghoondoodoom/peer-market/blob/d3842bbf7000f19764c0bfd1a40eb7e5fbb4f629/src%2Fmain%2Fjava%2Fdoodoom%2Fproject%2Fpeermarket%2Fcontroller%2FMemberController.java)|100%|:green_apple:|
|[MemberRegisterInput.java](https://github.com/younghoondoodoom/peer-market/blob/d3842bbf7000f19764c0bfd1a40eb7e5fbb4f629/src%2Fmain%2Fjava%2Fdoodoom%2Fproject%2Fpeermarket%2Fdto%2FMemberRegisterInput.java)|77.12%|:x:|
|[Member.java](https://github.com/younghoondoodoom/peer-market/blob/d3842bbf7000f19764c0bfd1a40eb7e5fbb4f629/src%2Fmain%2Fjava%2Fdoodoom%2Fproject%2Fpeermarket%2Fdomain%2FMember.java)|71.97%|:x:|
|[GlobalExceptionHandler.java](https://github.com/younghoondoodoom/peer-market/blob/d3842bbf7000f19764c0bfd1a40eb7e5fbb4f629/src%2Fmain%2Fjava%2Fdoodoom%2Fproject%2Fpeermarket%2Fexception%2FGlobalExceptionHandler.java)|9.72%|:x:|
|[ErrorResponse.java](https://github.com/younghoondoodoom/peer-market/blob/d3842bbf7000f19764c0bfd1a40eb7e5fbb4f629/src%2Fmain%2Fjava%2Fdoodoom%2Fproject%2Fpeermarket%2Fdto%2FErrorResponse.java)|0%|:x:|

|Total Project Coverage|73.07%|:x:|
|:-|:-:|:-:|

**Comment by younghoondoodoom** [comment]
|File|Coverage [100%]|:green_apple:|
|:-|:-:|:-:|
|[MemberServiceImpl.java](https://github.com/younghoondoodoom/peer-market/blob/6d0d18a8a672873aa0ae75d492e1b11410cf14c9/src%2Fmain%2Fjava%2Fdoodoom%2Fproject%2Fpeermarket%2Fservice%2Fmember%2FMemberServiceImpl.java)|100%|:green_apple:|
|[ErrorCode.java](https://github.com/younghoondoodoom/peer-market/blob/6d0d18a8a672873aa0ae75d492e1b11410cf14c9/src%2Fmain%2Fjava%2Fdoodoom%2Fproject%2Fpeermarket%2Ftype%2FErrorCode.java)|100%|:green_apple:|
|[MemberStatus.java](https://github.com/younghoondoodoom/peer-market/blob/6d0d18a8a672873aa0ae75d492e1b11410cf14c9/src%2Fmain%2Fjava%2Fdoodoom%2Fproject%2Fpeermarket%2Ftype%2FMemberStatus.java)|100%|:green_apple:|
|[BaseTimeEntity.java](https://github.com/younghoondoodoom/peer-market/blob/6d0d18a8a672873aa0ae75d492e1b11410cf14c9/src%2Fmain%2Fjava%2Fdoodoom%2Fproject%2Fpeermarket%2Fdomain%2Fbase%2FBaseTimeEntity.java)|100%|:green_apple:|
|[MemberController.java](https://github.com/younghoondoodoom/peer-market/blob/6d0d18a8a672873aa0ae75d492e1b11410cf14c9/src%2Fmain%2Fjava%2Fdoodoom%2Fproject%2Fpeermarket%2Fcontroller%2FMemberController.java)|100%|:green_apple:|
|[CustomAuthenticationFailureHandler.java](https://github.com/younghoondoodoom/peer-market/blob/6d0d18a8a672873aa0ae75d492e1b11410cf14c9/src%2Fmain%2Fjava%2Fdoodoom%2Fproject%2Fpeermarket%2Fconfiguration%2FCustomAuthenticationFailureHandler.java)|100%|:green_apple:|

|Total Project Coverage|100%|:green_apple:|
|:-|:-:|:-:|

**Comment by younghoondoodoom** [comment]
|File|Coverage [100%]|:green_apple:|
|:-|:-:|:-:|
|[MemberServiceImpl.java](https://github.com/younghoondoodoom/peer-market/blob/6d0d18a8a672873aa0ae75d492e1b11410cf14c9/src%2Fmain%2Fjava%2Fdoodoom%2Fproject%2Fpeermarket%2Fservice%2Fmember%2FMemberServiceImpl.java)|100%|:green_apple:|
|[ErrorCode.java](https://github.com/younghoondoodoom/peer-market/blob/6d0d18a8a672873aa0ae75d492e1b11410cf14c9/src%2Fmain%2Fjava%2Fdoodoom%2Fproject%2Fpeermarket%2Ftype%2FErrorCode.java)|100%|:green_apple:|
|[MemberStatus.java](https://github.com/younghoondoodoom/peer-market/blob/6d0d18a8a672873aa0ae75d492e1b11410cf14c9/src%2Fmain%2Fjava%2Fdoodoom%2Fproject%2Fpeermarket%2Ftype%2FMemberStatus.java)|100%|:green_apple:|
|[BaseTimeEntity.java](https://github.com/younghoondoodoom/peer-market/blob/6d0d18a8a672873aa0ae75d492e1b11410cf14c9/src%2Fmain%2Fjava%2Fdoodoom%2Fproject%2Fpeermarket%2Fdomain%2Fbase%2FBaseTimeEntity.java)|100%|:green_apple:|
|[MemberController.java](https://github.com/younghoondoodoom/peer-market/blob/6d0d18a8a672873aa0ae75d492e1b11410cf14c9/src%2Fmain%2Fjava%2Fdoodoom%2Fproject%2Fpeermarket%2Fcontroller%2FMemberController.java)|100%|:green_apple:|
|[CustomAuthenticationFailureHandler.java](https://github.com/younghoondoodoom/peer-market/blob/6d0d18a8a672873aa0ae75d492e1b11410cf14c9/src%2Fmain%2Fjava%2Fdoodoom%2Fproject%2Fpeermarket%2Fconfiguration%2FCustomAuthenticationFailureHandler.java)|100%|:green_apple:|

|Total Project Coverage|100%|:green_apple:|
|:-|:-:|:-:|

File: src/main/java/doodoom/project/peermarket/exception/MemberException.java [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+import doodoom.project.peermarket.type.ErrorCode;
+import lombok.Getter;
+
+@Getter
+public class MemberException extends RuntimeException {
+    private final ErrorCode errorCode;
+    private final String errorMessage;
+
+    public MemberException(ErrorCode errorCode) {
```
Comment (spring-kotlin-lover) [inline src/main/java/doodoom/project/peermarket/exception/MemberException.java]: MemberException 내부에 errorCode 를 함축하고 있는 구조네요.

커스텀에 대한 예외로 정의해주신 부분은 좋았습니다. 다만 합성되는 구조가 조금 특이한데요.
MemberException 을 사용하는 개발자 관점에서 보자면 ErrorCode 라는 합성구조가 조금 어색한 것 같아요.

ErrorCode 는 추상화된 일반적인 ErrorCode 를 의미하고, MemberException 은 구체적인 클래스로 한정해서 사용하게 되는데,
MemberException 을 정의할때 다시 또 추상화된 ErrorCode 를 주입해주어야 하니 사용하는 사람 입장에서 정의하기 어려운 점이 있는 것 같습니다.

그렇다고 한다면, 레벨을 맞추어야 하는 것을 추천드리는데요. 한번 깊이 있게 고민해보시면 좋겠네요 😄

**Review by spring-kotlin-lover — COMMENTED** [review]
안녕하세요. 저는 무지멘토이구요~! 작성해주신 코드보니까 전반적으로 공부를 조금 해보셨던 것 같아 깊이 있는 주제로 코멘트 하나 드렸습니다.

File: src/main/java/doodoom/project/peermarket/exception/MemberException.java (hunk 1 above)
Comment (younghoondoodoom) [reply src/main/java/doodoom/project/peermarket/exception/MemberException.java]: 안녕하세요 멘토님! 얼마 전까지 개인적인 사정으로 인해 프로젝트 진행이 안되었습니다ㅠㅠ 확인은 했었는데 이제야 제대로 고민해보네요. 죄송합니다🥲 
해주신 피드백 너무 감사드립니다! 혼자는 고민해보지 못한 너무 좋은 고민 거리이네요!

고민해보니 MemberException같은 구체적인 클래스가 추상화된 ErrorCode에 의존하는게 이상하네요. 
여러가지를 고민해봤는데 그냥 단순하게 MembereException을 여러개로 쪼개는게 나을 것 같습니다. MemberException 하나로 여러가지 예외를 처리하려고 하다보니까 더 복잡해지는 느낌이네요.. 

레벨을 맞추어야 한다는게 어떤 기준의 레벨인지 잘 모르겠지만 일단 다시 커밋하겠습니다. 피드백 부탁드립니다 😀

Comment (younghoondoodoom) [reply src/main/java/doodoom/project/peermarket/exception/MemberException.java:11]: @spring-kotlin-lover 멘토님, 레벨을 맞추어야한 다는 것이 무슨 뜻인지 잘 모르겠습니다. MemberException 안에 에러 코드와 메세지를 포함 시켜서 static field로 만들어야 한다는 뜻일까요?

**Comment by younghoondoodoom** [comment]
|File|Coverage [100%]|:green_apple:|
|:-|:-:|:-:|
|[MemberServiceImpl.java](https://github.com/younghoondoodoom/peer-market/blob/28920ccd5be0097be1bdafff25a8a9657d9707e0/src%2Fmain%2Fjava%2Fdoodoom%2Fproject%2Fpeermarket%2Fservice%2Fmember%2FMemberServiceImpl.java)|100%|:green_apple:|
|[MemberStatus.java](https://github.com/younghoondoodoom/peer-market/blob/28920ccd5be0097be1bdafff25a8a9657d9707e0/src%2Fmain%2Fjava%2Fdoodoom%2Fproject%2Fpeermarket%2Ftype%2FMemberStatus.java)|100%|:green_apple:|
|[BaseTimeEntity.java](https://github.com/younghoondoodoom/peer-market/blob/28920ccd5be0097be1bdafff25a8a9657d9707e0/src%2Fmain%2Fjava%2Fdoodoom%2Fproject%2Fpeermarket%2Fdomain%2Fbase%2FBaseTimeEntity.java)|100%|:green_apple:|
|[MemberController.java](https://github.com/younghoondoodoom/peer-market/blob/28920ccd5be0097be1bdafff25a8a9657d9707e0/src%2Fmain%2Fjava%2Fdoodoom%2Fproject%2Fpeermarket%2Fcontroller%2FMemberController.java)|100%|:green_apple:|
|[CustomAuthenticationFailureHandler.java](https://github.com/younghoondoodoom/peer-market/blob/28920ccd5be0097be1bdafff25a8a9657d9707e0/src%2Fmain%2Fjava%2Fdoodoom%2Fproject%2Fpeermarket%2Fconfiguration%2FCustomAuthenticationFailureHandler.java)|100%|:green_apple:|

|Total Project Coverage|100%|:green_apple:|
|:-|:-:|:-:|

File: src/main/java/doodoom/project/peermarket/service/member/MemberServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+public class MemberServiceImpl implements MemberService {
+
+    private final MemberRepository memberRepository;
+    private final PasswordEncoder passwordEncoder;
+
+    @Override
+    @Transactional
+    public boolean register(MemberRegisterInput input) {
+        if (memberRepository.findByEmail(input.getEmail()).isPresent()) {
+            throw new MemberException(EMAIL_ALREADY_EXIST);
```
Comment (lyhjohn) [inline src/main/java/doodoom/project/peermarket/service/member/MemberServiceImpl.java]: 이메일이 중복일 경우 에러페이지로 리턴하기 보다 다시 회원가입 form으로 리턴해주는 방식은 어떠실까요?

File: src/main/java/doodoom/project/peermarket/service/member/MemberServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+                .isAdmin(false)
+                .build();
+        memberRepository.save(member);
+        return true;
+    }
+
+    @Override
+    public UserDetails loadUserByUsername(String email) throws UsernameNotFoundException {
+        Member member = memberRepository.findByEmail(email).orElseThrow(
+                () -> new UsernameNotFoundException(MEMBER_NOT_FOUND.getDescription()));
```
Comment (lyhjohn) [inline src/main/java/doodoom/project/peermarket/service/member/MemberServiceImpl.java]: UsernameNotFoundException 예외를 던진 후 이를 처리하는 로직이 필요할 것 같습니다!

File: src/main/java/doodoom/project/peermarket/type/ErrorCode.java [hunk 4]
```diff
@@ -0,0 +1,15 @@
+package doodoom.project.peermarket.type;
+
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+
+@Getter
+@AllArgsConstructor
+public enum ErrorCode {
```
Comment (lyhjohn) [inline src/main/java/doodoom/project/peermarket/type/ErrorCode.java]: 단순 ErrorCode 파일 하나에 전부 넣는 것 보다는 RegisterErrorCode, LoginErrorCode 이런 식으로 나눠보시는 건 어떨까요? (개인적인 의견이라 정답이 없을 수 있습니다!)

**Review by lyhjohn — COMMENTED** [review]
고생 많으셨습니다!

File: src/main/java/doodoom/project/peermarket/service/member/MemberServiceImpl.java (hunk 2 above)
Comment (younghoondoodoom) [reply src/main/java/doodoom/project/peermarket/service/member/MemberServiceImpl.java]: 이번에는 예외를 일괄적으로 처리하고 싶어서 그 방법은 피해봤습니다! 물론 그 방식이 ux가 더 좋겠지만 예외처리를 일괄적으로 하려고 한 선택입니다. 그리고 개인적으로 view 파일에 시간을 오래 투자하고 싶지 않았습니다.. 피드백 감사드립니다😃

File: src/main/java/doodoom/project/peermarket/service/member/MemberServiceImpl.java (hunk 3 above)
Comment (younghoondoodoom) [reply src/main/java/doodoom/project/peermarket/service/member/MemberServiceImpl.java]: UsernameNotFoundException이 났다는 것을 사용자한테 알리고싶지 않았습니다! 그냥 loginForm으로 redirect만 하는 전략을 사용했습니다!

File: src/main/java/doodoom/project/peermarket/type/ErrorCode.java (hunk 4 above)
Comment (younghoondoodoom) [reply src/main/java/doodoom/project/peermarket/type/ErrorCode.java]: 이 부분은 enum을 사용하지 않겠다는 전략을 선택했고 현재 수정 후 커밋했습니다. 피드백 감사드립니다!

File: src/main/java/doodoom/project/peermarket/exception/MemberException.java (hunk 1 above)
Comment (spring-kotlin-lover) [reply src/main/java/doodoom/project/peermarket/exception/MemberException.java]: 지금 작성해주신 방향성도 좋네요 ㅎㅎ 일반화하기 어렵다면 쪼개는 것도 괜찮은 전략입니다.

**Pull request merged**